src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubReceiver.java

```diff
@@ -0,0 +1,546 @@
+ * the License.
+ */
+package io.cdap.plugin.gcp.publisher.source;
+
+import com.google.api.gax.core.FixedCredentialsProvider;
+import com.google.api.gax.rpc.ApiException;
+import com.google.api.gax.rpc.StatusCode;
+import com.google.auth.Credentials;
+import com.google.cloud.pubsub.v1.SubscriptionAdminClient;
+import com.google.cloud.pubsub.v1.stub.GrpcSubscriberStub;
+import com.google.cloud.pubsub.v1.stub.SubscriberStub;
+import com.google.cloud.pubsub.v1.stub.SubscriberStubSettings;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.pubsub.v1.AcknowledgeRequest;
+import com.google.pubsub.v1.ProjectSubscriptionName;
+import com.google.pubsub.v1.PullRequest;
+import com.google.pubsub.v1.PullResponse;
+import com.google.pubsub.v1.PushConfig;
+import com.google.pubsub.v1.ReceivedMessage;
+import com.google.pubsub.v1.TopicName;
+import io.cdap.plugin.gcp.common.GCPUtils;
+import org.apache.spark.storage.StorageLevel;
+import org.apache.spark.streaming.receiver.Receiver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+/**
+ * Spark Receiver for Pub/Sub Messages.
+ * <p>
+ * If backpressure is enabled, the message ingestion rate for this receiver will be managed by Spark.
+ */
+public class PubSubReceiver extends Receiver<PubSubMessage> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PubSubReceiver.class);
+  private static final String CREATE_SUBSCRIPTION_ERROR_MSG =
+    "Failed to create subscription '%s'.";
+  private static final String CREATE_SUBSCRIPTION_ADMIN_CLIENT_ERROR_MSG =
+    "Failed to create subscription client to manage subscription '%s'.";
+  private static final String CREATE_SUBSCRIPTION_RETRY_ERROR_MSG =
+    "Failed to create subscription '%s' after 5 attempts";
+  private static final String MISSING_TOPIC_ERROR_MSG =
+    "Failed to create subscription. Topic '%s' was not found in project '%s'.";
+  private static final String SUBSCRIBER_ERROR_MSG =
+    "Failed to create subscriber using subscription '%s' for project '%s'.";
+  private static final String FETCH_ERROR_MSG =
+    "Failed to fetch new messages using subscription '%s' for project '%s'.";
+
+  // Retryable status codes. These need to be handled in case Pub/Sub throws a StatusRuntimeException.
+  private static final int RESOURCE_EXHAUSTED = 429;
+  private static final int CANCELLED = 499;
+  private static final int INTERNAL = 500;
+  private static final int UNAVAILABLE = 503;
+  private static final int DEADLINE_EXCEEDED = 504;
```

Comment:
These are standard HTTP code. Use the HttpURLConnection.HTTP_* constants instead.

Reply:
Not all of them appear in the `HttpURLConnection` class.

Would it help to use Google's Status Codes? https://googleapis.dev/java/gax/1.44.0/com/google/api/gax/rpc/StatusCode.Code.html